File: intersection-of-two-linked-lists/intersection-of-two-linked-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *getIntersectionNode(ListNode *head1, ListNode *head2) {
        
        ListNode* temp1 = head1;
        ListNode* temp2 = head2;
        unordered_set<ListNode*> st;

        while(temp1!=NULL & temp2!=NULL){
            if(st.find(temp1)!=st.end()){
                return temp1;
            }
            st.insert(temp1);
            temp1 = temp1->next;
            if(st.find(temp2)!=st.end()){
                return temp2;
            }
            st.insert(temp2);
            temp2 = temp2->next;
        }
        
        while(temp2!=NULL){
           if(st.find(temp2)!=st.end()){
               return temp2;
           }
           temp2 = temp2->next;
        }

        while(temp1!=NULL){
           if(st.find(temp1)!=st.end()){
               return temp1;
           }
           temp1 = temp1->next;
        }
        return NULL;
    }
};

File: 160-intersection-of-two-linked-lists/intersection-of-two-linked-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:

    def findIntersectionNode(self, root):

        node1 = root
        node2 = root

        while(node2!=None and node2.next !=None):
            node1 = node1.next
            node2 = node2.next.next
            if node1 == node2:
                if node1 == None:
                    return None
                break

        while(root!=node1 and node1!=None):
            root = root.next
            node1 = node1.next
        
        return node1


    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> Optional[ListNode]:
        
        endA = headA

        while(endA.next!=None):
            endA = endA.next

        endA.next = headB

        ans = self.findIntersectionNode(headA)
        endA.next = None
        
        return ans